go/bundle/cmd/sign-bundle/main.go

```diff
@@ -0,0 +1,133 @@
+package main
+
+import (
+	"crypto"
+	"flag"
+	"io/ioutil"
+	"log"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/WICG/webpackage/go/bundle"
+	"github.com/WICG/webpackage/go/bundle/signature"
+	"github.com/WICG/webpackage/go/signedexchange"
+	"github.com/WICG/webpackage/go/signedexchange/certurl"
+)
+
+var (
+	flagInput        = flag.String("i", "in.wbn", "Webbundle input file")
+	flagOutput       = flag.String("o", "out.wbn", "Webbundle output file")
+	flagCertificate  = flag.String("certificate", "cert.cbor", "Certificate chain CBOR file")
+	flagPrivateKey   = flag.String("privateKey", "cert-key.pem", "Private key PEM file")
+	flagValidityUrl  = flag.String("validityUrl", "https://example.com/resource.validity.msg", "The URL where resource validity info is hosted at.")
+	flagDate         = flag.String("date", "", "Datetime for the signature in RFC3339 format (2006-01-02T15:04:05Z). (default: current time)")
+	flagExpire       = flag.Duration("expire", 1*time.Hour, "Validity duration of the signature")
+	flagMIRecordSize = flag.Int("miRecordSize", 4096, "Record size of Merkle Integrity Content Encoding")
+)
+
+func readCertChainFromFile(path string) (certurl.CertChain, error) {
+	fi, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer fi.Close()
+	return certurl.ReadCertChain(fi)
+}
+
+func readPrivateKeyFromFile(path string) (crypto.PrivateKey, error) {
+	privkeytext, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	return signedexchange.ParsePrivateKey(privkeytext)
+}
+
+func readBundleFromFile(path string) (*bundle.Bundle, error) {
+	fi, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer fi.Close()
+	return bundle.Read(fi)
+}
+
+func writeBundleToFile(b *bundle.Bundle, path string) error {
+	fo, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return err
+	}
+	defer fo.Close()
+	_, err = b.WriteTo(fo)
+	return err
+}
+
+func addSignature(b *bundle.Bundle, signer *signature.Signer) error {
+	for _, e := range b.Exchanges {
+		if !signer.CanSignForURL(e.Request.URL) {
+			continue
+		}
+		payloadIntegrityHeader, err := e.AddPayloadIntegrity(b.Version, *flagMIRecordSize)
+		if err != nil {
+			return err
+		}
+		if err := signer.AddExchange(e, payloadIntegrityHeader); err != nil {
+			return err
+		}
+	}
+
+	newSignatures, err := signer.UpdateSignatures(b.Signatures)
+	if err != nil {
+		return err
+	}
+	b.Signatures = newSignatures
+	return nil
+}
+
+func main() {
```

Comment:
Optional nit: we may want to copy main()/run() idiom: https://github.com/WICG/webpackage/blob/master/go/bundle/cmd/dump-bundle/main.go#L62

Reply:
Done.